rdflib/namespace/__init__.py

```diff
@@ -176,6 +169,59 @@ def __repr__(self):
         return "URIPattern(%r)" % str(self)
 
 
+class DefinedNamespaceMeta(type):
+    """
+    Utility metaclass for generating URIRefs with a common prefix
+
+    """
+
+    _NS: Namespace
+    _warn: bool = True
+    _fail: bool = False             # True means mimic ClosedNamespace
+    _extras: List[str] = []         # List of non-pythonesque items
+    _underscore_num: bool = False    # True means pass "_n" constructs
+
+    def __getitem__(cls, name, default=None):
+        name = str(name)
+        if str(name).startswith("__"):
+            return super().__getitem__(name, default)
+        if (cls._warn or cls._fail) and not name in cls:
+            if cls._fail:
+                raise AttributeError(f"term '{name}' not in namespace '{cls._NS}'")
+            else:
+                warnings.warn(f"Code: {name} is not defined in namespace {cls.__name__}", stacklevel=3)
+        return cls._NS[name]
+
+    def __getattr__(cls, name):
+        return cls.__getitem__(name)
+
+    def __repr__(cls):
+        return f'Namespace("{cls._NS}")'
+
+    def __str__(cls):
+        return str(cls._NS)
+
+    def __contains__(cls, item):
+        """ Determine whether a URI or an individual item belongs to this namespace """
+        item_str = str(item)
+        if item_str.startswith("__"):
+            return super().__contains__(item)
+        if item_str.startswith(str(cls._NS)):
+            item_str = item_str[len(str(cls._NS)):]
+        return any(item_str in c.__annotations__ or item_str in c._extras or
+                   (cls._underscore_num and item_str[0] == '_' and item_str[1:].isdigit())
+                   for c in cls.mro() if issubclass(c, DefinedNamespace))
+
+
+class DefinedNamespace(metaclass=DefinedNamespaceMeta):
```

Comment:
I haven't touched ClosedNamespace, so yes.

DefinedNamespace's behavior is sufficiently different that it is easy (and safer, imo) to implement it as a wrapper.

Reply:
Also, in response to the generator question above, yes - it was built using https://github.com/hsolbrig/definednamespace -- once (if) the above changes are accepted in rdflib then I can do a cleanup step.  It is broken right at the moment and needs a good deal more documentation.

It leads to a question, however.  I could drop the generator into the tools directory if it made sense or leave it as a separate package.  Will be a bit 'o a pita to get all the unit tests and the like migrated so it may be a little while...

Reply:
I think getting it into the primary tools directory of this repository is a good idea. A more extreme view would be that it is preferable for any code generation components to be packaged with the core, and thus that inclusion be a perquisite for merging this PR, but that seems silly to me since the generation code can be merged at its own pace. Let's see what others think.